File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/GatrixClient.ts
/**
 * GatrixClient - Main entry point for Gatrix Client SDK
 *
 * Provides access to various Gatrix services:
 * - features: Feature flags
 * - (future) surveys: Surveys
 * - (future) maintenance: Maintenance status
 */
import { EventEmitter } from './EventEmitter';
import { GatrixClientConfig, GatrixSdkStats } from './types';
import { FeaturesClient } from './FeaturesClient';
import { EVENTS } from './events';
import { SDK_VERSION } from './version';
import { GatrixError } from './errors';
import { Logger, ConsoleLogger } from './Logger';

export class GatrixClient {
  private emitter: EventEmitter;
  private config: GatrixClientConfig;
  private featuresClient: FeaturesClient;
  private initialized = false;
  private startPromise: Promise<void> | null = null;
  private logger: Logger;

  /**
   * Feature flags client
   * Access via client.features.isEnabled(), client.features.boolVariation(), etc.
   */
  get features(): FeaturesClient {
    return this.featuresClient;
  }

  constructor(config: GatrixClientConfig) {
    // Validate required config
    if (!config.apiUrl) {
      throw new GatrixError('apiUrl is required');
    }
    if (!config.apiToken) {
      throw new GatrixError('apiToken is required');
    }
    if (!config.appName) {
      throw new GatrixError('appName is required');
    }
    if (!config.environment) {
      throw new GatrixError('environment is required');
    }

    this.config = config;
    this.emitter = new EventEmitter();
    this.logger = config.logger ?? new ConsoleLogger('GatrixClient');
    this.featuresClient = new FeaturesClient(this.emitter, config);
  }

  /**
   * Start the SDK
   * Initializes all services and begins polling for updates
   */
  async start(): Promise<void> {
    if (this.startPromise) {
      return this.startPromise;
    }

    if (this.initialized) {
      return Promise.resolve();
    }

    const connId = this.featuresClient.getConnectionId();
    this.logger.info(`Starting SDK for ${this.config.appName} (v${SDK_VERSION}) [${connId}]`);

    this.startPromise = (async () => {
      try {
        await this.featuresClient.init();
        await this.featuresClient.start();
        this.initialized = true;
      } finally {
        this.startPromise = null;
      }
    })();

    return this.startPromise;
  }

  /**
   * Stop the SDK
   * Stops all polling and cleanup
   */
  stop(): void {
    this.featuresClient.stop();
    this.initialized = false;
    this.startPromise = null;
  }

  /**
   * Check if SDK is ready
   */
  isReady(): boolean {
    return this.featuresClient.isReady();
  }

  /**
   * Get last error
   */
  getError(): unknown {
    return this.featuresClient.getError();
  }

  /**
   * Get SDK statistics (combined from all services)
   */
  getStats(): GatrixSdkStats {
    const featStats = this.featuresClient.getStats();
    return {
      sdkState: featStats.sdkState || (this.getError() ? 'error' : this.isReady() ? 'healthy' : 'initializing'),
      startTime: featStats.startTime || null,
      connectionId: this.featuresClient.getConnectionId(),
      errorCount: featStats.errorCount ?? 0,
      lastError: featStats.lastError ?? this.getError(),
      lastErrorTime: featStats.lastErrorTime || null,
      offlineMode: this.config.offlineMode ?? false,
      features: featStats,
      eventHandlerStats: this.emitter.getHandlerStats(),
    };
  }

  // ==================== Event Subscription ====================

  /**
   * Subscribe to an event
   */
  on(event: string, callback: (...args: any[]) => void, name?: string): this {
    this.emitter.on(event, callback, name);
    return this;
  }

  /**
   * Subscribe to an event once
   */
  once(event: string, callback: (...args: any[]) => void, name?: string): this {
    this.emitter.once(event, callback, name);
    return this;
  }

  /**
   * Unsubscribe from an event
   */
  off(event: string, callback?: (...args: any[]) => void): this {
    this.emitter.off(event, callback);
    return this;
  }

  /**
   * Subscribe to ALL events
   * Callback receives (eventName, ...args)
   */
  onAny(callback: (event: string, ...args: any[]) => void, name?: string): this {
    this.emitter.onAny(callback, name);
    return this;
  }

  /**
   * Unsubscribe from ALL events listener
   */
  offAny(callback?: (event: string, ...args: any[]) => void): this {
    this.emitter.offAny(callback);
    return this;
  }

  // ==================== Static Helpers ====================

  /**
   * Get SDK version
   */
  static get version(): string {
    return SDK_VERSION;
  }

  /**
   * Event constants
   */
  static get EVENTS() {
    return EVENTS;
  }
}

File: packages/sdks/client-sdks/gatrix-unity-sdk/Runtime/GatrixClient.cs
// GatrixClient - Main entry point for Gatrix Unity SDK
// Top-level class wrapping FeaturesClient and other services

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;

namespace Gatrix.Unity.SDK
{
    /// <summary>
    /// Main entry point for the Gatrix Unity SDK.
    /// Wraps FeaturesClient and provides high-level API.
    /// </summary>
    public class GatrixClient : IDisposable
    {
        public static string SdkName => SdkInfo.Name;
        public static string SdkVersion => SdkInfo.Version;

        private readonly GatrixEventEmitter _emitter;
        private readonly GatrixClientConfig _config;
        private readonly HttpClient _httpClient;
        private FeaturesClient _featuresClient;
        private bool _started;
        private bool _disposed;

        public GatrixClient(GatrixClientConfig config)
        {
            ValidateConfig(config);

            _config = config;
            _emitter = new GatrixEventEmitter();
            _httpClient = new HttpClient();

            // Add SDK identification header
            _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(
                "X-SDK-Name", SdkName);
            _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(
                "X-SDK-Version", SdkVersion);

            _featuresClient = new FeaturesClient(_emitter, _config, _httpClient);
        }

        /// <summary>Get the FeaturesClient for direct access</summary>
        public FeaturesClient Features => _featuresClient;

        /// <summary>Get the event emitter for subscribing to events</summary>
        public GatrixEventEmitter Events => _emitter;

        /// <summary>Check if SDK is ready</summary>
        public bool IsReady => _featuresClient?.IsReady() ?? false;

        /// <summary>Check if SDK has started</summary>
        public bool IsStarted => _started;

        /// <summary>Get the connection ID</summary>
        public string ConnectionId => _featuresClient?.GetConnectionId();

        // ==================== Lifecycle ====================

        /// <summary>
        /// Initialize and start the SDK.
        /// Call this once during application startup.
        /// </summary>
        public async ValueTask StartAsync()
        {
            if (_started) return;

            await _featuresClient.InitAsync();
            await _featuresClient.StartAsync();
            _started = true;
        }

        /// <summary>Stop the SDK (cancels polling, stops metrics)</summary>
        public void Stop()
        {
            _featuresClient?.Stop();
            _started = false;
        }

        // ==================== Event Subscription ====================

        /// <summary>Subscribe to an SDK event</summary>
        public GatrixClient On(string eventName, GatrixEventHandler callback, string name = null)
        {
            _emitter.On(eventName, callback, name);
            return this;
        }

        /// <summary>Subscribe to an SDK event once</summary>
        public GatrixClient Once(string eventName, GatrixEventHandler callback, string name = null)
        {
            _emitter.Once(eventName, callback, name);
            return this;
        }

        /// <summary>Unsubscribe from an SDK event</summary>
        public GatrixClient Off(string eventName, GatrixEventHandler callback = null)
        {
            _emitter.Off(eventName, callback);
            return this;
        }

        /// <summary>Subscribe to ALL SDK events</summary>
        public GatrixClient OnAny(GatrixAnyEventHandler callback, string name = null)
        {
            _emitter.OnAny(callback, name);
            return this;
        }

        /// <summary>Unsubscribe from ALL SDK events listener</summary>
        public GatrixClient OffAny(GatrixAnyEventHandler callback = null)
        {
            _emitter.OffAny(callback);
            return this;
        }

        // ==================== Statistics ====================

        /// <summary>Get SDK statistics (combined from all services)</summary>
        public GatrixSdkStats GetStats()
        {
            var featStats = _featuresClient.GetStats();
            return new GatrixSdkStats
            {
                SdkState = featStats.SdkState,
                StartTime = featStats.StartTime,
                ConnectionId = _featuresClient.GetConnectionId(),
                ErrorCount = featStats.ErrorCount,
                LastError = featStats.LastError,
                LastErrorTime = featStats.LastErrorTime,
                OfflineMode = _config.OfflineMode,
                Features = featStats,
                EventHandlerStats = _emitter.GetHandlerStats()
            };
        }

        // ==================== Dispose ====================

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            Stop();
            _emitter.RemoveAllListeners();
            _httpClient?.Dispose();
        }

        // ==================== Validation ====================

        private static void ValidateConfig(GatrixClientConfig config)
        {
            if (config == null)
                throw new ArgumentNullException(nameof(config));
            if (string.IsNullOrEmpty(config.ApiUrl))
                throw new ArgumentException("apiUrl is required", nameof(config));
            if (string.IsNullOrEmpty(config.ApiToken))
                throw new ArgumentException("apiToken is required", nameof(config));
            if (string.IsNullOrEmpty(config.AppName))
                throw new ArgumentException("appName is required", nameof(config));
            if (string.IsNullOrEmpty(config.Environment))
                throw new ArgumentException("environment is required", nameof(config));
        }
    }
}
